content/docs/using-pulumi/debugging/_index.md

```diff
@@ -0,0 +1,285 @@
+```
+
+Then, prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```py
+import debugpy
+
+debugpy.listen(("localhost", 5678))
+debugpy.wait_for_client()
+
+# ... Your normal Pulumi code follows here
+```
+
+Run `pulumi up` the usual way:
+
+```bash
+pulumi up
+Previewing update (dev)
+
+ Type                 Name  Plan   Info
+ pulumi:pulumi:Stack  dev          0.00s - Note: Debugging will proceed.
+```
+
+Set a breakpoint in your Pulumi code.
+
+Navigate to your IDE and attach the debugger to the process with the `python` runtime. For example, in VS Code, you can define the following `launch.json` file:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Python: Local Attach",
+            "type": "debugpy",
+            "request": "attach",
+            "connect": {
+                "host": "localhost",
+                "port": 5678
+            },
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}",
+                    "remoteRoot": "."
+                }
+            ],
+            "justMyCode": true
+        }
+    ]
+}
+```
+
+navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Python: Local Attach` configuration.
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging .NET programs
+
+Prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```cs
+using System.Diagnostics;
+
+return await Pulumi.Deployment.RunAsync(async () =>
+{
+    while (!Debugger.IsAttached)
+    {
+        await System.Threading.Tasks.Task.Delay(1000);
+    }
+    Debugger.Break();
+
+    // ... Your normal Pulumi code follows here
+});
+```
+
+Run `pulumi up` the usual way:
+
+```bash
+pulumi up
+Previewing update (dev)
+
+ Type                 Name          Plan   Info
+ pulumi:pulumi:Stack  play-cs-dev          'dotnet build -nologo .' completed successfully
+```
+
+Navigate to your IDE and attach the debugger to the process with the name of your .NET project. For example, in VS Code, you can define the following `launch.json` file:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": ".NET Core Attach",
+            "type": "coreclr",
+            "request": "attach"
+        }
+    ]
+}
+```
+
+navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `.NET Core Attach` configuration. Search for the process by your project name and attach:
+
+![Debug: Attach to .NET Process](./dotnet-code-attach.png)
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging Go programs
+
+Prepend your Pulumi program with the code to pause and give yourself time to attach the debugger:
```

Comment:
I think you should mention that the Go program should be complied with debug flags using `go build -gcflags all=-N -l`, and then use the `runtime.binary` field in the project file to launch the built executable.

Reply:
> Please mention that the Go program should be compiled with debug flags using go build -gcflags all=-N -l, then used as the binary in Pulumi.yaml. Otherwise breakpoints may not work as expected.

I did not need to change anything about the compilation flags, it just worked. Was it just luck?

Reply:
I think it _almost_ works without flags, but can be frustrating. [Here's](https://www.jetbrains.com/help/go/attach-to-running-go-processes-with-debugger.html#step-1-build-the-application) an example of the routine guidance.